lib/rbac/query_shared_resource.rb

```diff
@@ -0,0 +1,64 @@
+require 'rbac-api-client'
+module RBAC
+  class QuerySharedResource
+    include Utilities
+    attr_accessor :share_info
+
+    def initialize(options)
+      @app_name = options[:app_name]
+      @resource_id = options[:resource_id]
+      @resource_name = options[:resource_name]
+      @share_info = []
+      verbs = options[:verbs] || []
+      @regexp = if verbs.present?
+                  Regexp.new("#{@app_name}:#{@resource_name}:(#{verbs.join('|')})")
+                else
+                  Regexp.new("#{@app_name}:#{@resource_name}:")
+                end
+      @verb_regexp = Regexp.new("#{@app_name}:#{@resource_name}:(?<verb>.*)")
+    end
+
+    def process
+      roles_from_policies
+      self
+    end
+
+    private
+
+    def roles_from_policies
+      RBAC::Service.call(RBACApiClient::PolicyApi) do |api_instance|
+        RBAC::Service.paginate(api_instance, :list_policies, {}).each do |item|
+          filter_roles(item.roles, item.group)
```

Comment:
@syncrou 
The  filter logic is different for unshare and querying. During unshare we have to keep the role objects around since we might update it. During query we don't have to store the roles we just collect what we need in the share_info array.